keps/sig-node/4960-container-stop-signals/README.md

```diff
@@ -0,0 +1,471 @@
+// ...
+-	if err := c.stopContainer(ctx, container, time.Duration(r.GetTimeout())*time.Second); err != nil {
++ 	if err := c.stopContainer(ctx, container, time.Duration(r.GetTimeout())*time.Second, container.Config.Lifecycle.StopSignal); err != nil {
+		return nil, err
+	}
+// ...
+}
+
+-func (c *criService) stopContainer(ctx context.Context, container containerstore.Container, timeout time.Duration) error {
++func (c *criService) stopContainer(ctx context.Context, container containerstore.Container, timeout time.Duration, containerStopSignal string) error {
+//...
+    if timeout > 0 {
+	    stopSignal := "SIGTERM"
+-       if container.StopSignal != "" {
++       if containerStopSignal != "" {
++   	    stopSignal = containerStopSignal 
++       } else if  container.StopSignal != "" {
+			stopSignal = container.StopSignal
+		} else {
+// rest of the code...
+```
+
+The signal that we get from `ContainerConfig` can be validated with [ParseSignal](https://github.com/containerd/containerd/blob/main/vendor/github.com/moby/sys/signal/signal.go#L38) to further validate that we've received a valid stop signal. Also here `container.StopSignal` is reading the stop signal from the image. We can add another condition before that to use the stop signal defined in spec if there is one. If nothing is defined in the spec ("" or unset), containerd behaves like how it is today. Also note that `SIGTERM` is hardcoded in containerd's stopContainer method as the default stop signal to fallback to, in case the image doesn't defined a stop signal. Similar logic in also present in CRI-O [here](https://github.com/cri-o/cri-o/blob/main/internal/oci/container.go#L259-L272).
+
+Find the entire diff for containerd which was done for the POC [here](https://github.com/containerd/containerd/compare/main...sreeram-venkitesh:containerd:added-custom-stop-signal?expand=1).
+
+### User Stories (Optional)
+
+#### Story 1
+
+Kubernetes by default sends a SIGTERM to all containers while killing them. When running nginx on Kubernetes, this can result in nginx dropping requests as reported [here](https://github.com/Kong/kubernetes-ingress-controller/pull/283). The current solution for this issue would be to build custom images with a SIGQUIT stop signal or to write a PreStop lifecycle hook that manually kills the process gracefully, which is what is done in the PR. If we had stop signal support at the Container spec level, this would've been easier and straightforward to implement. Users wouldn't have to patch the applications running on Kubernetes to handle different termination behavior. This is also similar to [this issue](https://github.com/github/resque/pull/21). 
+
+### Risks and Mitigations
+
+- We'll be adding the complexity of signal handling to the pod/container spec. If users define an signal that is not handled by their containers, this can lead to pods hanging. 
```

Comment:
by default, the terminationGracePeriodSeconds is 30 seconds. For the default terminationGracePeriodSeconds setting, pod kill will hang for just 30 seconds by default. 

https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#lifecycle

Do I understand correctly?

Reply:
Yes this is correct, I have updated this in the KEP so that it is clearer.